### PR TITLE
#28: Added chunked Parquet reader

### LINKED ISSUE #28: Add chunked Parquet reader
### Situation

Parquet files are organized into row groups. These chunks of a single parquet file then can be read instead of reading the whole file. The PIOJ should support reading specified row groups.

### Acceptance Criteria

- Chunked Parquet reader is implemented
- In addition, Parquet file virtual splitter is provided
- Tests are added

### PR DESCRIPTION
**Pull request by morazow**
Fixes #28

### CODE REVIEW

File: src/main/java/com/exasol/parquetio/splitter/ParquetFileSplitter.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        var startPosition = 0L;
+        var currentSize = 0L;
+        for (var endPosition = 0; endPosition < end; endPosition++) {
+            currentSize += rowGroups.get(endPosition).getTotalByteSize();
+            if (currentSize >= this.chunkSize) {
+                chunks.add(new ChunkInterval(startPosition, endPosition + 1L));
+                startPosition = endPosition + 1L;
+                currentSize = 0L;
+            }
+        }
```
Comment (jakobbraun) [inline src/main/java/com/exasol/parquetio/splitter/ParquetFileSplitter.java:72]: This operation looks expensive. Have you checked how long the planning takes? Maybe assuming that all row groups are equally sized would be an acceptable trade off.

Comment (jakobbraun) [reply src/main/java/com/exasol/parquetio/splitter/ParquetFileSplitter.java:72]: Or we could split the file size into partitions and determine which block belong into a partition during reading.

File: src/test/java/com/exasol/parquetio/reader/RowParquetChunkReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        final var reader = new RowParquetChunkReader(ParquetTestFileWriter.getInputFile(path), chunks);
+        assertValues(reader);
+    }
+
+    private void assertValues(final RowParquetChunkReader reader) {
+        List<Integer> values = collectValues(reader);
+        for (int value = 0; value <= 4000; value++) {
+            assertThat(values.contains(value), equalTo(true));
+        }
+        for (int value = 4200; value < RECORD_COUNT - 2500; value++) {
```
Comment (jakobbraun) [inline src/test/java/com/exasol/parquetio/reader/RowParquetChunkReaderTest.java:50]: What's with the gab from 4000 to 4200?

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/parquetio/data/Interval.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.exasol.parquetio.data;
+
+/**
+ * An interval interface that holds start and end positions {@code [...)} of each splitted chunk.
+ */
+public interface Interval {
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/main/java/com/exasol/parquetio/data/Interval.java:6]: Yes, I had interval with filename in mind.

File: src/test/java/com/exasol/parquetio/reader/RowParquetChunkReaderTest.java (hunk 2 above)
Comment (morazow) [reply src/test/java/com/exasol/parquetio/reader/RowParquetChunkReaderTest.java:74]: Yes, that is rough estimate, it is numbers in single row group with 8KB size. I was not sure if exact number will be same in different environments. I will check what was the exact, maybe they are same.

File: src/main/java/com/exasol/parquetio/splitter/ParquetFileSplitter.java (hunk 1 above)
Comment (morazow) [reply src/main/java/com/exasol/parquetio/splitter/ParquetFileSplitter.java:72]: No it is not expensive. Number of row groups, their sizes are already written with file at the footer. It only reads the available metadata from file itself.

They are mostly equally sized, by default ~128MB, but users can set while writing using `.withRowGroupSize(rowGroupSize)`.

File: src/main/java/com/exasol/parquetio/splitter/FileSplitter.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.exasol.parquetio.splitter;
+
+import java.util.List;
+
+import com.exasol.parquetio.data.Interval;
+
+/**
+ * An interface for splitting files into chunks.
+ */
+public interface FileSplitter {
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/main/java/com/exasol/parquetio/splitter/FileSplitter.java:10]: I'll keep it for now. Maybe we may need different splitters or file partitioners.

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**